File: leetcode/035-search-insert-position/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int lo = 0, hi = (int)nums.size();
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] < target) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 3, 5, 6};
    cout << sol.searchInsert(nums, 5) << endl; // 2
    cout << sol.searchInsert(nums, 2) << endl; // 1
    cout << sol.searchInsert(nums, 7) << endl; // 4
    return 0;
}

File: leetcode/035-search-insert-position/solution.py
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums)
        while lo < hi:
            mid = (lo + hi) // 2
            if nums[mid] < target:
                lo = mid + 1
            else:
                hi = mid
        return lo


if __name__ == "__main__":
    print(Solution().searchInsert([1, 3, 5, 6], 5))  # 2
    print(Solution().searchInsert([1, 3, 5, 6], 2))  # 1
    print(Solution().searchInsert([1, 3, 5, 6], 7))  # 4
